### ISSUE TITLE
Updating to MBGL iOS SDK 6.3.0 on M1 chipset causes cocoapods linking error

### ISSUE THREAD

**Issue by naftalibeder** [react-native-mapbox-gl/maps#1354]
Has anyone had any success upgrading to 6.3.0 on iOS, e.g.

`$ReactNativeMapboxGLIOSVersion = '6.3.0'`

When I switch this value from 6.2.1, and do a clean cocoapods install, it produces the following compile error:

> library not found for -lPods-MyProject

which ultimately I can trace down to the `Link Binary With Libraries` build phase. The only linked library in the list is titled `libPods-MyProject.a`, which references a file in `Pods/Products/` that doesn't exist. If I remove that link, or just remove the entire build phase, the project compiles and runs properly.

But this doesn't seem like the right solution, because doing a clean `pod install` automatically generates that build phase, so to fix this I would need to override pretty standard default behavior.

The same process, using 6.2.1, actually generates the `libPods-MyProject.a` file. I'm not sure where the bug actually lies. Does anyone have any ideas?

**Comment by gokugen** [comment]
Hi,
I think that you have forgotten to add this in your podfile:
```js
  pre_install do |installer|
    $RNMBGL.pre_install(installer)
  end
  
  post_install do |installer|
    $RNMBGL.post_install(installer)
  end

  $ReactNativeMapboxGLIOSVersion = '~> 6.3.0'
```

**Comment by naftalibeder** [comment]
@gokugen Thanks for the response. Unfortunately it still produces the same error. I've tested whittling down the Podfile to its essentials, and I still can't get around the error:

```
platform :ios, '10.0'

target 'MyProject' do
  config = use_native_modules!
    
  pre_install do |installer|
    # Mapbox
    $RNMBGL.pre_install(installer)
  end

  # React Native
  use_react_native!(:path => config["reactNativePath"])

  # Mapbox
  $ReactNativeMapboxGLIOSVersion = '~> 6.3.0'

  target 'MyProjectTests' do
    inherit! :search_paths
  end
  
  post_install do |installer|
    # Mapbox
    $RNMBGL.post_install(installer)
  end
end
```

Any other ideas?

**Comment by ferdicus** [comment]
can you post a few more specs, that might help others to debug:
* XCode Version
* Cocoapod version
* this repo version
* RN version

**Comment by naftalibeder** [comment]
Good call, my bad!

```
Xcode: Version 12.5 (12E262)
Cocoapods: 1.10.1
RNMBGL: 8.2.0-beta2 (current master)
React Native: 0.63.4
```

**Comment by naftalibeder** [comment]
Hi, does anyone have any thoughts?

**Comment by ferdicus** [comment]
I can only vouch for:
```
Xcode: Version 12.5 (12E262)
Cocoapods: 1.10.1
RNMBGL: 8.2.0-beta2 (current master)
React Native: 0.64.1
```

And it's building fine for my job.

I can try on friday with the `/example` and let you know more

**Comment by ferdicus** [comment]
A week late but yeah - `/example` builds fine with the above mentioned configuration

**Comment by naftalibeder** [comment]
Thank you for the response, no worries on the delay. I've isolated the problem into [a small repo](https://github.com/TruckMap/mapbox-examples/tree/upgrade-ios-sdk). Does this reproduce for you?

Any other ideas? The one thing I could think of is that I'm on an M1 Mac, but that doesn't explain why the previous version of the SDK worked fine.

**Comment by ferdicus** [comment]
Hey @naftalibeder, 
I've checked out [your repo](https://github.com/TruckMap/mapbox-examples/tree/upgrade-ios-sdk) and did the following:
* add `accessToken` in `index.js`
* adjust `    "@react-native-mapbox-gl/maps": "git@github.com:react-native-mapbox-gl/maps.git#master"` => `    "@react-native-mapbox-gl/maps": "git://github.com/react-native-mapbox-gl/maps.git#master"` (to be on `master`)
* ran `yarn`
* got an error with the post-install script
```sh
$ cd ios && bundle install && bundle exec pod install && cd ..
Traceback (most recent call last):
	2: from /usr/bin/bundle:23:in `<main>'
	1: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems.rb:302:in `activate_bin_path'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems.rb:283:in `find_spec_for_exe': Could not find 'bundler' (2.2.20) required by your /Users/user/development/mapbox-examples/ios/Gemfile.lock. (Gem::GemNotFoundException)
To update to the latest version installed on your system, run `bundle update --bundler`.
To install the missing version, run `gem install bundler:2.2.20`
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```
* executed `cd ios && pod install`
* after successful cocoapodding went back into root and ran
* `yarn react-native run-ios` with the following output
```sh
> $ yarn react-native run-ios
yarn run v1.22.5
$ /Users/user/development/mapbox-examples/node_modules/.bin/react-native run-ios
error React Native CLI uses autolinking for native dependencies, but the following modules are linked manually:
  - @react-native-mapbox-gl/maps (to unlink run: "react-native unlink @react-native-mapbox-gl/maps")
This is likely happening when upgrading React Native from below 0.60 to 0.60 or above. 
Going forward, you can unlink this dependency via "react-native unlink <dependency>" and it will be included in your app automatically. 
If a library isn't compatible with autolinking, disregard this message and notify the library maintainers.
Read more about autolinking: https://github.com/react-native-community/cli/blob/master/docs/autolinking.md
info Found Xcode workspace "MapPadding.xcworkspace"
info Building (using "xcodebuild -workspace MapPadding.xcworkspace -configuration Debug -scheme MapPadding -destination id=<some-id>")
success Successfully built the app
info Installing "/Users/user/Library/Developer/Xcode/DerivedData/MapPadding-effhljcihbvtzngdllpzksbhwxhu/Build/Products/Debug-iphonesimulator/MapPadding.app"
info Launching "org.reactjs.native.example.MapPadding"
success Successfully launched the app on the simulator
✨  Done in 7.31s.
```

not sure about the linking message 🤷🏿 

anyways, I'm greeted by this afterwards:
![Simulator Screen Shot - iPhone 12 - 2021-06-18 at 12 07 29](https://user-images.githubusercontent.com/26439946/122547147-040e8480-d030-11eb-93f5-798ca2b71f4a.png)


So all seems to work - at least on simulator?

You have seen my XCode specs above - plus, I do _not_ run an `M1`. 

Have you tried building this on another machine without `M1` chipset?

**Comment by naftalibeder** [comment]
Thank you so much for testing, and I'm sorry for the delay. I've had some colleagues attempt the same install, and sure enough, it fails on devices with an M1, and succeeds on those without.

For the purpose of leaving this issue so that I and people with the same problem can take some action, I am looking for an automated way to remove the `libPods-MyProject.a` library from the `Link Binary With Libraries` section, ideally in the `Podfile` itself. (This will allow development with the new version, before the iOS SDK actually supports the M1.)

Does anybody have any tips on how to achieve this? I've looked around and haven't been able to pin it down.

**Comment by ferdicus** [comment]
the mapbox iOS SDK will most probably never support M1 chips as the pre-v10 versions are no longer supported by mapbox.  
I would suggest you have a look into the fork from `MapLibre` (which we now also support) and give that a chance.  

You can check this repos issues to find mentions of M1 support and linked upstream tickets, one of them points to MapLibre and how they fixed a specific issue over there. 

worth a try

**Comment by naftalibeder** [comment]
I appreciate the tip, the MapLibre fork seems to work great!

I'm wondering what the long-term plan for this library is - do you expect to go all-in on MapLibre, or eventually migrate to the v10 version of the iOS SDK? Both seem to have pursued Metal support, but I'm not sure which is a better bet overall.

**Comment by ferdicus** [comment]
Hey @naftalibeder, I think it makes sense to move this lib to point to MapLibre as the default pack-in SDK:
* it's open source 
* same api that our react-native wrapper already supports

It's just an assessment from my side. 
As there are in essence only two maintainers of this repo (@mfazekas  and myself) I think it's the most pragmatic approach.  

 I wouldn't have the time (or skill, for that matter) to make this lib support v10 of mapbox. All responsibility would lie on one persons shoulders, which, considering "bus factor" and all.... I'm not sure makes that much sense.  

Long story short, we have to make a decision: 
* support 3 SDKs semi well/ bad (old mapbox SDKs/ v10 SDKs/ MapLibre SDKs) 
*  support 1 SDK more or less well (MapLibre)

Still open for discussion, however I don't see any other roadmap without more maintainers committing to helping out

**Comment by ferdicus** [comment]
btw, feel free to close this ticket :)

**Comment by naftalibeder** [comment]
For what it's worth, we did a release pointing this library to MapLibre.

On Android, it seems to have helped significantly, doing away with a number of frequent, obscure crashes, and increasing our crash-free rate from ~99% to ~99.8%. However, on iOS it went very poorly, dropping our crash-free rate from ~99.9% all the way down to ~88% (!).

So for now, we are going to split platforms, keeping MapLibre on Android but reverting to the original Mapbox SDK on iOS. I've also posted an issue on the [MapLibre](https://github.com/maplibre/maplibre-gl-native/issues/123) fork.

All of this is to say that I would be in favor of this library supporting both for the moment, with the idea that it incrementally will move towards sold a supporting MapLibre. How does that sound? I'll close this ticket but feel free to respond.

**Comment by ferdicus** [comment]
Glad it helped at one front, sad it degraded on the other :( 

We're not planning to immediately drop Mapbox support, if that's the fear. 
At the moment MapLibre and ("old") Mapbox are fairly similar.   

We'd probably rather opt to not pursue v10 of Mapbox and stick with MapLibre instead for future versions. 

This all obviously depends on contributors/ time 🤷🏿
